sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1372,15 +1374,16 @@ object CombineUnions extends Rule[LogicalPlan] {
         // Push down projection through Union and then push pushed plan to Stack if
         // there is a Project.
         case Project(projectList, Distinct(u @ Union(children, byName, allowMissingCol)))
-            if projectList.forall(_.deterministic) && children.nonEmpty &&
+            if projectList.forall(e => e.deterministic && !hasSubquery(e)) && children.nonEmpty &&
```

Comment:
I don't think we need to check subquery here. The code in Spark 3.2 can pushdown project though union as well even if it contains subqueries.

Reply:
fixed